File: python/basics/quicksort.py
#!/usr/bin/env python3
# Sort a fixed list with a quicksort over the head and tail of the list.


def quicksort(items):
    if len(items) <= 1:
        return list(items)

    pivot, *rest = items
    smaller = [x for x in rest if x <= pivot]
    larger = [x for x in rest if x > pivot]
    return quicksort(smaller) + [pivot] + quicksort(larger)


print(" ".join(str(value) for value in quicksort([5, 3, 8, 4, 2, 7, 1, 10, 9, 6])))

File: perl/basics/quicksort.pl
#!/usr/bin/env perl
# Sort a fixed list with a quicksort over the head and tail of the argument list.

use strict;
use warnings;

sub quicksort {
    return @_ if @_ <= 1;

    my ( $pivot, @rest ) = @_;
    my @smaller = grep { $_ <= $pivot } @rest;
    my @larger  = grep { $_ > $pivot } @rest;
    return ( quicksort(@smaller), $pivot, quicksort(@larger) );
}

print join( ' ', quicksort( 5, 3, 8, 4, 2, 7, 1, 10, 9, 6 ) ), "\n";
